src/librustc/infer/mod.rs

```diff
@@ -480,16 +480,16 @@ impl<'a, 'tcx> TyCtxt<'a, 'tcx, 'tcx> {
     {
         assert!(!value.needs_subst());
         let value = self.erase_late_bound_regions(value);
-        self.normalize_associated_type(&value)
+        self.normalize_associated_type_in(&value)
     }
 
     /// Fully normalizes any associated types in `value`, using an
     /// empty environment and `Reveal::All` mode (therefore, suitable
     /// only for monomorphized code during trans, basically).
-    pub fn normalize_associated_type<T>(self, value: &T) -> T
+    pub fn normalize_associated_type_in<T>(self, value: &T) -> T
```

Comment:
Nit: can we call this `normalize_associated_types_in` (i.e., pluralize `type`)

Reply:
I checked the codebase and there already exists a function named this way, [here](https://github.com/rust-lang/rust/blob/835e3e50781af375e6fef2d17e1bfa50d4e4aba3/src/librustc_typeck/check/mod.rs#L651). Would it be a problem if we had 2 different functions with the same name? (Mainly for people reading both of them).

Reply:
Ideally only one of them would remain as the public API, if I'm not mistaken.

Reply:
Hmm. So really these functions serve two distinct roles:

- The existing `normalize_associated_types_in` (implemented on `Inherited`) are used during type-checking, and hence may have a non-empty environment and/or inference variables. That particular one partly exists to silently register the extra obligations that comes from normalization.
- The `normalize_associated_type` function in contrast is intended to be used after monomorphization, and it forces an empty environment and `Reveal::All`. In this mode, we don't expect extra obligations to matter to the user, so we handle them internally. 

I think its pretty clear that the "naming convention" distinguishing the two is not sufficient. \</sarcasm\>

In other parts of the code, we've used the `trans_` prefix to distinguish the case of an empty environment with `Reveal::All` -- or at least code that runs after type-checking. @eddyb doesn't like that for the -- quite correct! -- reason that the code is generally applicable beyond trans. For example, MIR optimizations may want to use these pathways (though those may not have an empty environment, so that's slightly different). 

I think a lot of this comes down to whether you are running "before" or "after" typeck -- in particular, after typeck you know that errors should not occur, which can simplify the code and function signatures in various ways. So perhaps we can call this new function `normalize_associated_types_after_typeck`? We could also rename the existing `trans_` things in a similar vein.

Reply:
Oh if one of them is only in `rustc_typeck` then I got confused a bit, it certainly *feels* like the name shows up in a few places.

How about we just propagate the `ParamEnv` in more places and just make sure that in `rustc_trans` we give a `Reveal::All`? That seems simpler and less headache-inducing.

Reply:
@eddyb the signature currently is `T -> T`. I agree with changing it to `(ParamEnv, T) -> T` but that is not enough to remove the need for two functions. In particular, there are also "auxiliary obligations" and we have to decide what to do with them. The `FnCtxt` version ("during typeck") enrolls them into the fulfillment context. This version we are editing here solves them, assering that there are no errors (because it runs after type-check). The unifier form would probably propagate them back but nobody really wants that.

Reply:
Right, then, just take a `Reveal` and don't put `trans` in the name?

Reply:
@eddyb I don't see how that solves the problem. First off, taking a parameter environment -- not just reveal -- seems like the right thing to do (and is orthogonal from how we handle subobligations). Second, even if we just took a reveal, I still think we want a naming convention to distinguish these two cases ("expects to run after typeck" vs "does not").

Reply:
One option would be to use a combinator approach. Something like `tcx.post_typeck().normalize_associated_types_in(...)`?